core/modules/helpers/helpers.lua

```diff
@@ -0,0 +1,52 @@
+local set = require('set')
+local array = require('array')
+
+local helpers = {}
+
+
+---
+-- Finds all include directories required for the configuration/project,
+-- following project links to gather public include directories for those
+-- projects as well.
+---
+
+function helpers.fetchAllIncludeDirs(configOrProject)
+	-- Determine if we're working on a project level or a configuration level.
+	-- The logic is pretty much the same, but for configurations we'll look up
+	-- the same named configuration on the linked project.
+	local isProject = configOrProject.project == nil
+
+	local includeDirs = set.join(configOrProject.includeDirs.public, configOrProject.includeDirs.private)
+
+	-- Use an array of projects to visit plus a set of projects we've visited
```

Comment:
Should we do some kind of prevalidation to enforce an acyclic graph?

Reply:
That'd probably be good if it's not already in there somewhere. I still think this is useful here for non-acyclic graphs in the case that you have something like A references B and C, and B also references C. Otherwise you'll visit C twice and duplicate include directories. I think the comment here could probably be improved to mention that since it's a good reason to have this set logic even if we guarantee acyclic graphs.